stdlib/LinearAlgebra/src/diagonal.jl

```diff
@@ -380,13 +340,59 @@ function rdiv!(A::AbstractMatrix, D::Diagonal)
     A
 end
 
-function rdiv!(A::Union{LowerTriangular,UpperTriangular}, D::Diagonal)
-    broadcast!(/, parent(A), parent(A), permutedims(D.diag))
-    A
+(\)(D::Diagonal, A::AbstractMatrix) =
+    ldiv!(D, (typeof(oneunit(eltype(D))/oneunit(eltype(A)))).(A))
+(\)(D::Diagonal, b::AbstractVector) = D.diag .\ b
+(\)(Da::Diagonal, Db::Diagonal) = Diagonal(Da.diag .\ Db.diag)
+
+ldiv!(x::AbstractVecOrMat, A::Diagonal, b::AbstractVecOrMat) = (x .= A.diag .\ b)
```

Comment:
This one doesn't throw a `SingularException`. I think we have a fallback for 3-arg `ldiv!`, something like `ldiv!(copyto!(x, b), A)` or so, that admittedly has two passes over `x`, but otherwise would inherit all the exception handling from the function below. Not sure what's the right balance between code duplication and performance.

Reply:
I just noticed that this is not a new method, just moved here... Still, I think we should decide whether we want to remove this method and let it be handled by the generic fallback
```julia
# Generic fallback. This assumes that B and Y have the same sizes.
ldiv!(Y::AbstractArray, A::AbstractMatrix, B::AbstractArray) = ldiv!(A, copyto!(Y, B))
```
or have a fully-featured 3-arg version. The fallback seems to indicate we prefer to have 2-arg `ldiv!`s.

Reply:
2-arg version is preferred in general fallback because we call LAPACK to handle `ldiv!` in most cases?
For these broadcast based methods, 3-arg version seems to be more general (and efficient)? 
So maybe we can put the kernel in 3-arg version, and call it like `ldiv!(D::Diagnal, x) = ldiv!(x,D,x) `

BTW, since we have 3-arg `ldiv!`, does it make sense to add   a 3-arg `rdiv!`?

Reply:
I agree. Let's make the 3-arg `ldiv!` the default for `Digaonal`s and define 2-arg `ldiv!` as you say. It seems like we don't have any 3-arg `rdiv!` methods available, but 47 2-arg `rdiv!`s. So let's stick to that. The 3-arg `ldiv!` has to do with that for some factorizations one can only handle 3-arg versions, so we have a generic fallback, as well, IIRC.

Reply:
Done.
I think all `\` `/` related methods in think file do needed check, include singular and size, properly via `ldiv!` and `rdiv!`.